Shared wallets in incomplete state should not undergo restoration

paweljakubas opened the pull request:
### Issue Number

ADP-2432

### Description

- [x] I have made sure `Incomplete` shared wallet invokes  `createNonRestoringWalletWorker`.
- [x] When `Incomplete -> Ready` transition then the private key and meta from incomplete wallet is read.
- [x] The incomplete wallet is deleted and active wallet is created. The wallet starts restoring. 
- [x] Wallet takes meta and private key from incomplete wallet.

The solution making db exchange between workers (non-restoring and restoring) did not work properly. hence the above solution.

paweljakubas committed:
more on registering

paweljakubas committed:
introduce createNonrestoringWalletWorker

paweljakubas committed:
use createNonrestoringWalletWorker in post endpoints

paweljakubas committed:
try deleting worker

paweljakubas committed:
some simplifications

paweljakubas committed:
solution - part 1

paweljakubas committed:
deal with wallet name

paweljakubas committed:
use attach private key using prvx and shelley-schemed hashed pwd

paweljakubas committed:
handle account based wallets

paweljakubas committed:
reuse meta from pending wallet in active wallet

jonathanknowles committed:
Tweak formatting of function `postSharedWalletFromRootXPrv`.

jonathanknowles committed:
Tweak formatting of function `postSharedWallet`.

jonathanknowles committed:
Tweak formatting of function `postSharedWalletFromAccountXPub`.

jonathanknowles committed:
Tweak formatting of function `mkSharedWallet`.

jonathanknowles committed:
Tweak formatting of function `patchSharedWallet`.

jonathanknowles committed:
Use function `void` instead of assignment to wildcard.

jonathanknowles commented on lib/wallet/api/http/Cardano/Wallet/Api/Http/Server/Error.hs:
```diff
@@ -769,6 +769,11 @@ instance IsServerError ErrDerivePublicKey where
         ErrAddCosignerKeyNoSuchWallet e -> toServerError e
+        ErrAddCosignerKeyNoMeta ->
+            apiError err503 NoMeta $ mconcat
+                [ "No meta of wallet was found in database during migration"
```
@paweljakubas 

When you refer to "no meta of wallet", do you mean "no wallet metadata"?

paweljakubas replied:
yes

jonathanknowles committed:
Capitalise `Restoring` in `createNonRestoringWalletWorker`.

```patch
- createNonrestoringWalletWorker
+ createNonRestoringWalletWorker
```

jonathanknowles reviewed (commented):
Hi @paweljakubas 

Thanks for making this PR!

I've given it a brief look (though still need to look further).

I have one question about the use of terminology:

When you say **"no meta"**, are you referring to a case where the database has **no metadata** for a given wallet?

jonathanknowles committed:
Rename `NoMeta` error to `WalletMetadataNotFound`.

This commit makes it clearer that it was not possible to find the
**metadata** for a particular wallet.

In this particular context, the term "meta" is arguably a bit hard to
understand, unless the reader happens to be familiar with the internals
of cardano wallet.

jonathanknowles committed:
Revise error messages relating to `WalletMetadataNotFound`.

This commit revises the error messages relating to
`WalletMetadataNotFound`, for readability and flow.

piotr-iohk commented on lib/wallet/api/http/Cardano/Wallet/Api/Http/Server/Error.hs:
```diff
@@ -769,10 +769,12 @@ instance IsServerError ErrDerivePublicKey where
+            apiError err503 WalletMetadataNotFound $ T.unwords
+                [ "It was not possible to find any metadata for the given"
+                , "wallet within the database. This could be because the"
+                , "wallet has yet to become active after being in the pending"
```
```suggestion
                , "wallet has yet to become active after being in the incomplete"
```

piotr-iohk replied:
Q: what are circumstances for this error to occur? Is it possible to produce the test that incurs this error?

paweljakubas replied:
if something wrong with db ie. when during being in incomplete db is erased

piotr-iohk commented on lib/wallet/api/http/Cardano/Wallet/Api/Http/Shelley/Server.hs:
```diff
@@ -1156,13 +1193,56 @@ patchSharedWallet
     withWorkerCtx ctx wid liftE liftE $ \wrk -> do
         liftHandler $ W.updateCosigner wrk wid (liftKey accXPub) cosigner cred
+    wal' <- fst <$> getWallet ctx (mkSharedWallet @_ @s @k) (ApiT wid)
+    -- we switch on restoring only after pending -> active transition
```
```suggestion
    -- we switch on restoring only after incomplete -> active transition
```

paweljakubas replied:
done

piotr-iohk commented on specifications/api/swagger.yaml:
```diff
@@ -4165,6 +4165,22 @@ x-errNoRootKey: &errNoRootKey
+        wallet within the database.
+
+        May occur when a shared wallet has not yet become active after being in
+        the pending state.
```
```suggestion
        the incomplete state.
```

paweljakubas replied:
done

piotr-iohk reviewed (commented):
I have played a bit and it seems to work as expected. Also e2e tests pass on the branch.

I think that we could have some integration tests:
 - producing the introduced `WalletMetadataNotFound` if that is possible
 - creating a shared wallet, funding it, then deleting it and creating again "in steps", i.e. create + patch and making sure the balance is as expected.

paweljakubas committed:
replacing pending with incomplete

paweljakubas committed:
add integration test showing the case

jonathanknowles committed:
Arrange constructors of `ApiError` into alphabetical order.

jonathanknowles committed:
Revise wording of shared wallet error messages.

jonathanknowles committed:
Regenerate `ApiError.json` with a larger number of samples.

Some constructors for `ApiError`, when encoded to JSON, have an extra
`info` field, containing a JSON object.

In order to ensure we cover the encoding of these `info` objects (which
at the present time are generated relatively infrequently), I've
regenerated the `ApiError.json` golden file with a larger number of
samples. (200 samples, to be exact.)

I used the following method:

```
rm lib/wallet/test/data/Cardano/Wallet/Api/ApiError.json
CREATE_MISSING_GOLDEN=TRUE GOLDEN_SAMPLE_SIZE=200 cabal run cardano-wallet:test:unit -- --match "produces the same JSON"
```

In future, we could alter the generator of `ApiError` so that it
generates these richer errors with a higher frequency. Alternatively, as
we add more and more rich JSON errors, this will become unnecessary.
Regardless, regenerating the golden file is probably good enough for
now.

jonathanknowles committed:
Fix formatting in `Integration.Scenario.API.Shared.Transactions`.

Wherever practical:

- 80-column lines
- 4 space indents

piotr-iohk commented on lib/wallet/integration/src/Test/Integration/Scenario/API/Shared/Transactions.hs:
```diff
@@ -837,7 +909,74 @@ spec = describe "SHARED_TRANSACTIONS" $ do
+        let walShared2 = getFromResponse Prelude.id rPatch2
+
+        --checking that balances are the same as before deletion
+        eventually "balance is like before deletion" $ do
```
:+1:

piotr-iohk commented on specifications/api/swagger.yaml:
```diff
@@ -5208,6 +5224,13 @@ x-responsesErr404WalletNotFound: &responsesErr404WalletNotFound
 
+x-responsesErr503WalletMetadataNotFound: &responsesErr503WalletMetadataNotFound
+  503:
+    description: No meta
```
A bit more elaborate description would be nice. :)

piotr-iohk commented on test/e2e/spec/spec_helper.rb:
```diff
@@ -175,7 +175,7 @@ def patch_if_incomplete(wid, payment_patch, deleg_patch)
   when 403
     expect(d_upd).to be_correct_and_respond 403
-    expect(d_upd.parsed_response['code']).to eq 'shared_wallet_not_pending'
+    expect(d_upd.parsed_response['code']).to eq 'shared_wallet_active'
```
:+1:

piotr-iohk reviewed (approved):
LGTM.
